### DISCUSSION TITLE
如何将DesignAble生成的jsonSchema反显在DesignAble上？

### DISCUSSION THREAD

**Discussion by sorryljt** [alibaba/formily#1912] [Q&A 2.x] [answered]
#### 如题

通过DesignAble生成一个页面的jsonSchema，并存到数据库，希望通过DesignAble对已有的jsonSchema进行可视化编辑，请问如何将其反显呢？有无直接的API？或者Demo

**Comment by atzcl** [comment, marked as answer]
### 大概做以下两步即可
- 将 ` json schema` 对象递归创建成以 `TreeNode` 实例为节点的节点树
- 将这棵树设置到当前的 `Workspace` 的 `operation.tree` 的下级节点中

这里可以直接使用 `@designable/formily` 提供的 `transformToTreeNode` 方法

### 具体的代码

这里对应上面的第一步
https://github.com/alibaba/formily/blob/5d88e29a439c4966dcaa1a83056000a43797323c/designable/antd/playground/widgets/SchemaEditorWidget.tsx#L24

这里对应上面的第二步
https://github.com/alibaba/formily/blob/5d88e29a439c4966dcaa1a83056000a43797323c/designable/antd/playground/main.tsx#L97-L101

这里的 `ViewPanel ` 的 `render props` 部分需要到 `designable` 那边去看
 https://github.com/alibaba/designable/blob/4bf6aaed683b2d30c677125d5379b6cd1f6658d3/packages/react/src/panels/ViewPanel.tsx#L20

https://github.com/alibaba/designable/blob/4bf6aaed683b2d30c677125d5379b6cd1f6658d3/packages/react/src/panels/ViewPanel.tsx#L34-L37

https://github.com/alibaba/designable/blob/4bf6aaed683b2d30c677125d5379b6cd1f6658d3/packages/core/src/models/TreeNode.ts#L611-L635

**Reply by wangjie19910615** [reply]
请问我要初始化挂在DESIGNABLE的ComponentTreeWidget组件一直不成功

**Reply by wangjie19910615** [reply]
请问把已知的node设置到当前的 Workspace 的 operation.tree 的下级节点中, 具体要怎么做？

**Reply by sorryljt** [reply]
您好，按照您上面上的两步试了一下，请问在第二步，在`ComponentTreeWidget `组件中怎么使用呢？是需要自己扩展他的渲染规则吗

**Reply by atzcl** [reply]
> 您好，按照您上面上的两步试了一下，请问在第二步，在`ComponentTreeWidget `组件中怎么使用呢？是需要自己扩展他的渲染规则吗

`ComponentTreeWidget ` 的作用就是递归渲染 `tree`，渲染的组件来源就是通过 `components` 传入的列表，然后根据渲染的 `node.componentName` 来匹配命中需要渲染的组件，这里的处理跟 `formily` 是如何将 `json schema` 渲染成页面节点的处理基本一样

因为 `designable` 使用了 `@formily/reactive` 进行响应式开发，所以我们只需要修改 `operation.tree`，就使得 `ComponentTreeWidget` 内部进行相应的响应式渲染

**Reply by atzcl** [reply]
> 请问把已知的node设置到当前的 Workspace 的 operation.tree 的下级节点中, 具体要怎么做？

`const tree = useTree()`, 然后改就行了，具体有啥属性方法，你可以点进去看看 `TreeNode` 模型

**Reply by sorryljt** [reply]
> > 您好，按照您上面上的两步试了一下，请问在第二步，在`ComponentTreeWidget `组件中怎么使用呢？是需要自己扩展他的渲染规则吗
> 
> `ComponentTreeWidget ` 的作用就是递归渲染 `tree`，渲染的组件来源就是通过 `components` 传入的列表，然后根据渲染的 `node.componentName` 来匹配命中需要渲染的组件，这里的处理跟 `formily` 是如何将 `json schema` 渲染成页面节点的处理基本一样
> 
> 因为 `designable` 使用了 `@formily/reactive` 进行响应式开发，所以我们只需要修改 `operation.tree`，就使得 `ComponentTreeWidget` 内部进行相应的响应式渲染

多谢耐心解答，然后，“我们只需要修改 `operation.tree`”就是这一步不知道要怎么实现，在哪改呀？是有方法还是修改属性呢？就这一步没有看懂，辛苦大佬再给细说下，我好像就差这一步

**Reply by wangjie19910615** [reply]
再请教一个问题，我在WorkspacePanel修改tree，只有DESIGNABLE和MARKUP生效了，JSONTREE和PREVIEW中都没有生效

**Reply by sorryljt** [reply]
> 再请教一个问题，我在WorkspacePanel修改tree，只有DESIGNABLE和MARKUP生效了，JSONTREE和PREVIEW中都没有生效

你怎么实现的在`DESIGNABLE `上生效呢？康康代码~

**Reply by wangjie19910615** [reply]
tree.appendNode(node)

**Reply by wangjie19910615** [reply]
要用tree.from(newTree)

**Reply by sorryljt** [reply]
> 要用tree.from(newTree)

害...没搞定，这代码是写在哪里的呀？方便给个代码块吗

**Reply by atzcl** [reply]
> > 要用tree.from(newTree)
> 
> 害...没搞定，这代码是写在哪里的呀？方便给个代码块吗

这个代码写在你从后端请求回 `json schema` 数据处，以下是伪代码，你需要根据自己的业务进行调整
```
// 伪代码

const tree = useTree()

const jsonSchema = await requestGetJsonSchema()

const newTreeNode = transformToTreeNode(jsonSchema)

tree.form(newTreeNode)
```

**Reply by sorryljt** [reply]
好的，多谢，已解决，不过在现有`<Designer/>`页面，使用useTree报错，我们换了一种写法，如下
```js
  const currentWorkspace = (window['__DESINGER_WORKSPACE__']) ? window['__DESINGER_WORKSPACE__'] : engine.workbench.currentWorkspace;
  const tree = currentWorkspace?.operation?.tree;
  if(tree && jsonSchemaString){
    const newTreeNode = transformToTreeNode(JSON.parse(jsonSchemaString),FormParser)
    tree.from(newTreeNode)
    tree.takeSnapshot()
  }
```
我的项目是直接拉下来的git上playground的代码，总感觉这个页面框架在渲染顺序上有点问题，有时候使用个实例，用个hook啥的，还没有初始化，下面的Dom就执行了，只能想上面那样避开

**Comment by janryWang** [comment]
最新的palyground已经支持了数据回显逻辑，可以参考 https://github.com/alibaba/formily/blob/formily_next/designable/antd/playground/widgets/ActionsWidget.tsx

**Reply by sorryljt** [reply]
> 最新的palyground已经支持了数据回显逻辑，可以参考 https://github.com/alibaba/formily/blob/formily_next/designable/antd/playground/widgets/ActionsWidget.tsx

收到，期待文档

**Reply by janryWang** [reply]
https://v2.formilyjs.org/zh-CN/guide/form-builder

**Reply by sorryljt** [reply]
收到，多谢

**Comment by wangchenyi1996** [comment]
请问一下大佬，如果是多个Workspace工作区， 来回切换，然后点击节点，页面就卡起来了， 2个工作区都点不动，在任意工作区，重新拖动就正常了， 这是啥原因呀
